(PUP-3020) add puppet config delete functionality

ericldelaney opened the pull request:
Add ability to delete config settings from the puppet.conf file

jhelwig commented on lib/puppet/face/config.rb:
```diff
@@ -129,4 +129,45 @@
+
+  action(:delete) do
+    summary _("Delete a Puppet setting.")
+    #TRANSLATORS '--section' is a command line option and should not be translated
```
Was this meant to be above the notes section further down?

ericldelaney replied:
No, I had added the the option to the argument, but then realized later that it wasn't needed and removed it. 

Removed

ericldelaney commented:
messages:
Deleted setting from 'master' with value: 'wow = thing'
Warning: No setting found in configuration file for section user setting name wow


✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:35 $ cat /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf
[main]
wow = thing
crazy = thing
[master]
wow = thing
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:35 $ bundle exec puppet config set crazy thing --section user
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:36 $ cat /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf
[main]
wow = thing
crazy = thing
[master]
wow = thing
[user]
crazy = thing
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:36 $
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:36 $ bundle exec puppet config delete wow --section master
Deleted setting from 'master' with value: 'wow = thing'
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:36 $ bundle exec puppet config delete wow --section user
Warning: No setting found in configuration file for section user setting name wow
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:36 $ bundle exec puppet config delete crazy --section user
Deleted setting from 'user' with value: 'crazy = thing'
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:36 $
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:37 $ cat /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf
[main]
wow = thing
crazy = thing
[master]
[user]
✔ ~/work/my_repos/puppet [task/master/PUP-3020/add_config_delete_action L|⚑ 1]
14:37 $

jhelwig commented on lib/puppet/settings/ini_file.rb:
```diff
@@ -34,6 +34,13 @@ def append(line)
+  def delete(section, name)
+    entry = setting(section, name)
+    if entry
+      @lines.reject! { |l| l == entry }
```
Each line in `@lines` has a reference to the previous line, so we'll need to make sure that the line after the one we want to remove gets updated so it's no longer pointing to the removed line.

jhelwig replied:
For clarity: It's not causing a problem when only deleting a line, but would if deleting and adding within the same Manipulator as adding uses the `previous` to find where it should do certain things.

ericldelaney commented:
@jhelwig Updated

puppetcla commented:
CLA signed by all contributors.

kris-bosland commented on lib/puppet/face/config.rb:
```diff
@@ -129,4 +129,45 @@
+
+  action(:delete) do
+    summary _("Delete a Puppet setting.")
+    #TRANSLATORS '--section' is a command line option and should not be translated
```
Is this comment associated with the notes string or example string below, instead?

kris-bosland commented on lib/puppet/face/config.rb:
```diff
@@ -129,4 +129,44 @@
+    when_invoked do |name, options|
+
+      path = Puppet::FileSystem.pathname(Puppet.settings.which_configuration_file)
+      Puppet::FileSystem.touch(path)
```
Is this touch to create the file if it doesn't exist?  Should we be doing that?

ericldelaney replied:
@kris-bosland  I copied the basic methods from the set() section, assuming that it was all there for a reason.

jhelwig replied:
With `set`, I think it makes sense to create a non-existent file, but with `delete`, I don't think it makes as much sense to try to create the config file if it doesn't already exist.

ericldelaney commented:
@kris-bosland @jhelwig Updated.


$ bundle exec puppet config set wow thing --section main
$ bundle exec puppet config delete wow --section main
Deleted setting from 'main' with value: 'wow = thing'
$ bundle exec puppet config delete wow --section main
Warning: No setting found in configuration file for section 'main' setting name 'wow'
$ mv /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf.save
$ bundle exec puppet config delete wow --section main
Warning: The puppet.conf file does not exist /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf
$

jhelwig commented:
With the current way `delete` in `IniFile`, and the "`Deleted setting from '%{section_name}' with value: '%{setting_string}'`" message interact, any leading whitespace from the original config line will be preserved in the message. So we could get output like the following, depending on how the original file was indented:

```
Deleted setting from 'main' with value: '                wow = thing'
```

Is this intentional? Wondering if it might be better to limit what gets returned from `delete` to be only the value of the setting, instead of the stringified config file line.

kris-bosland commented:
![Groovy!][https://media2.giphy.com/media/BvHvlxhRBI0Ja/giphy.gif]

ericldelaney commented:
@jhelwig, yes it was intentional. By including the setting you can recreate the file if you had the command output with less effort than if you  did something like

Deleted setting 'wow' from section 'main' with value: 'thing'

kris-bosland commented:
I saw the leading whitespace in the output too, and thought it was a wash: pro, preserve the text that was removed for the user, con looks like poor formatting.  I don't think there is a clear winner.

ericldelaney commented:
@jhelwig 

How about this then:

$ bundle exec puppet config set '  wow' thing --section main
$ cat /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf
  wow = thing
[main]
$ bundle exec puppet config delete wow --section main
Deleted setting from 'main': 'wow = thing'


$ bundle exec puppet config set wow thing --section main
$ bundle exec puppet config delete wow --section main
Deleted setting from 'main': 'wow = thing'
$ bundle exec puppet config delete wow --section main
Warning: No setting found in configuration file for section 'main' setting name 'wow'
$ mv /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf.save
$ bundle exec puppet config delete wow --section main
Warning: The puppet.conf file does not exist /Users/eric.delaney/.puppetlabs/etc/puppet/puppet.conf
$

ericldelaney commented on lib/puppet/face/config.rb:
```diff
@@ -129,4 +129,49 @@
+    EOT
+
+    when_invoked do |name, options|
+      options[:section] = options[:section].to_s # If value was left as default - set to default string
```
Added this conversion of the section from a symbol to a string, to work with the change in (PUP-2868) which is also up for review

jhelwig replied:
Left https://github.com/puppetlabs/puppet/pull/6603#issuecomment-363245919 about updating how section names are handled, since it seems like we shouldn't need to worry about doing this everywhere.

jhelwig commented:
#6603 can remove the `.to_s` if/when it updates things to that it's not necessary.
